### DISCUSSION TITLE
Meta_Hours_4 - Breakup of Yugoslavia

### DISCUSSION THREAD

**Discussion by yangwao** [kodadot/nft-gallery#2007] [Meta-Hours]
- [previous meta hours 3](https://github.com/kodadot/nft-gallery/discussions/1710)
- Our recent code quality area was like the breakup of Yugoslavia
- Next photo is of Berry, somewhere in mountains, performing parkour.

| ![](https://upload.wikimedia.org/wikipedia/commons/thumb/a/a4/Breakup_of_Yugoslavia-TRY2.gif/300px-Breakup_of_Yugoslavia-TRY2.gif) | <img height=400px src="https://user-images.githubusercontent.com/5887929/150561409-8b905b70-a1e8-4f49-99a5-bf76fb073f77.jpg">

### To be added
- we have closed a few issues labelled by `v2.2` part of https://github.com/kodadot/nft-gallery/issues/1758

### Stargazing

Our recent stargazing from Github users on our github repository [kodadot/nft-gallery](https://github.com/kodadot/nft-gallery) -- feel free to click on the star in the top right, you will be part of graph 😍

![Screenshot_2022-01-31-18-12-49-54_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://user-images.githubusercontent.com/5887929/151840723-c8763c35-33b8-427e-b8c1-47f3dcd804fb.jpg)

### Usual new all-time high records

- My bet is due to new integrations and a more intense and faster pace of development, we are hitting new highs. 

- Notice that just at 25/1 early morning, when the screenshot was taken, in 4 days we are hitting 300 build minutes, which if this could continue linear, we should hit slightly above 2k build minutes over a month, so seems our tempo is getting consistent and on a steady rise.

![Screenshot 2022-01-25 at 10 16 50](https://user-images.githubusercontent.com/5887929/150947922-0901a153-e371-417b-baa2-caa2fb1d5eb5.png)
![Screenshot 2022-01-25 at 10 16 59](https://user-images.githubusercontent.com/5887929/150947927-b081703c-89be-4632-b247-34627f0b4feb.png)
![Screenshot 2022-01-25 at 10 17 08](https://user-images.githubusercontent.com/5887929/150947929-4d670d3b-1b50-48e1-b73e-702f9b9a1345.png)

### Pull Requests -- usually jazz 

- In the previous billing cycle we've been **happy for on average 100 Pull requests per monthly basis**. 
- A month later, **we are hitting 200 Pull requests** which is amazing, to be exact, 218 pull requests. I would thank every developer who contributes to the KodaDot and keep it going. 
- Every pull request is well rewarded in $KSM and you are building really great public good. Whenever you have your friends looking for a side hustle, they are more than welcome to sharpen their skill  

![image](https://user-images.githubusercontent.com/5887929/152141801-f1041cad-3972-4385-89fc-cafd2d3a513a.png)

### Speed improvements

- You've maybe already noticed, even before Meta Hours 4, that our loading times are being cut and no matter where are you on Internet, the site loads way faster and NFTs as well. Particular [Cloudflare Imageez](https://github.com/kodadot/nft-gallery/pull/2143) @vikiival can speak about this later in a more deep dive
- **Cleaning code** -- Cleaning code is focused fo new developers to not being disturbed with non-used components.  
- With recent code cleanups, we've found that our site builds incredibly faster. 
- **From usual times install dependencies from on average 4m**, with removing unused packages **we've managed to land around 170 seconds**, `12:10:44 PM: Done in 172.18s.` to be precise. 
- With various updates in packages, removing unused packages and components #2046 #2001 #2031 we are managing to generate our site at **each build under 150 seconds**, which actually took way over 240 seconds (4  minutes), sometimes even 360second (6 minutes).
- Most importantly we installed a few bots 
- like to Renovate to help us bump our packages to be up to date and mitigate obvious security concerns for our end-user experience.
- Prettier to format code
- [Build](https://github.com/kodadot/nft-gallery/blob/main/.github/workflows/build.yml) heads-up to @roiLeo which will check everytime if code builds. Yes we have Netlify for this as well, but for bump packages we wanted to save minutes and wanted prepare room for shipping higher quality code with fewer bugs

![image](https://user-images.githubusercontent.com/5887929/150780330-f0ca4585-82a8-4bf3-9606-441a31f8f23e.png)

### Flat Collection Fee

We've decided to flatten collection fee.
Now the fee for minting the collection was around $0.07.
Since the last time changed the fee we had:

Free Textile
Free Pinata (only paid for IPFS pins)

#### Current stack which we need to pay running costs:
- Cloudflare Images
- Cloudflare Workes
- Cloudflare Pages
- Pinata Pro
- Pinata throughput for dedicated virtual gateway
#### Upcoming 
 - Indexers
- Our development costs are getting higher so it's time to start to covering them early from platform fees

#### Reasoning
- The basic flat fee is $0.5 per minted collection
- For the NFT mint it's also $0.5 + $1 if the NFT is carbonless
 
- Change in fees structure [has been made since merge 1967](https://github.com/kodadot/nft-gallery/pull/1967) 

### Incentivized testing

- As you may know, our app is being used around 7-11k unique users on monthly averages with hitting around 300k pageviews monthly average. After realizing, we would definitely make sure that our app works flawlessly and has less bugs as possible.
- After this conclusion, I'm thinking we should have public bug bounty program (can be related incl security issues) to hunt down various glitches that can occur. 
- Goal of this program would raise awareness of some unwanted behaviour of app.
- Anytime you spot the bug, heads up to our github, use bug form and report and describe it, of course, most described and detailed reported issue would have ability to be solved as fast as possible and will be preferred to be figured out
- rewards could be varied from simple bugs starting at $50 up to $600 to most complex ones
- we would write some piece of text on the kodadot about this, maybe @cryptodamsky can formulate some great text and we can put it on landing page under button as separate page and mention this bug-bounty program at docs.kodadot.xyz as well
- I've found that I was thinking about it back then #1222 
- Plus we have an incentive to found a security bounty program #537
- Plus we can have a hall of fame of best bug reporters (leaderboard) to drive awareness of their past work https://github.com/kodadot/nft-gallery/issues/1440

### Mitigation of me being a bottleneck of development and becoming more replaceable
- As you may see in the picture below, I've taken some days off by writing our new version to automatize bounties payout plus some mental downtime, which was really needed for my wellbeing. Meanwhile caught up some cold and other digestion issues, which are tackled even now bit but it's in a good way. 
- I've come to the realization that I'm being actually a bottleneck of processing stuff. Thus I'm happy to start testing this, let's call bot Ephemeral and once will prove it's okay, I'll add spending limits for our frequent contributors who have to write rights and approve PRs. - Was thinking something along **100USD per PR, 300USD weekly spending limit per approver** and we will observe how it goes with a vision to extend limits. Probably the result would be that I would doesn't need to be present at all and help me to keep focusing on laying out future strategies for KodaDot and Metaprime which is very needed right now. 
- If everything goes well, I will gradually add a spending limit for selected approvers and we will see how it would go from here.

- On another side, **I'm really happy that our async workflow and backlogged issues really work well and have the non-blocking environment as much as possible tho, plus we are living really around the world**

<img height=800px src="https://user-images.githubusercontent.com/5887929/151692952-6fffd1a6-bb85-482b-a0e0-dda198a8574d.png">

### [docs.kodadot.xyz](https://docs.kodadot.xyz)

- I'm really happy that @cryptodamsky has stepped up the game and has started learning Git, Nodejs, Vuepress framework and will drive passion to keep up our documentation, maybe technical writer in future, who knows! Probably he will speak about his experience and learnings which was really something new for him. 
- I was really pleased to help him setup stuff, make onboarding call and now he is really great aspiring technical writer of KodaDot

![image](https://user-images.githubusercontent.com/5887929/151692994-30c7ea1b-39ad-4f9b-852c-94ff001f3a4e.png)

### What's we could done in this 
But we did not all of them.

- #2013
- #2065  
- #1758 
- #2076 
- #2077
- #2018 
- #1477 - we need leaderboard both for payouts for developers and payouts for bug bounty
- #1619 -- reflect what has been done since 27/12 @XyloDrone linked these issues 
- #1500 -- this was monsterous effort 
- #1966 we've renovate bot https://github.com/kodadot/nft-gallery/pull/1966

### Cherry-picked issues for future Meta Hours
- As you know we have a long list of backlogged issues, which sounds fine for internally or external suggestion, various brain dumps that might be duplicated and some issues that would require more attention than any other. I've come across our libraries of them and would be nice to dispatch a few of them for future Meta_Hours. 
- Like set it as potential achievement in our bi-weekly cycle

#### Comfort browsing
- https://github.com/kodadot/nft-gallery/issues/1357

#### Search
We have a few issues related to search as we would like to help people find what they are looking for much faster and more precise
- https://github.com/kodadot/nft-gallery/issues/1023
- https://github.com/kodadot/nft-gallery/issues/842
- https://github.com/kodadot/nft-gallery/issues/348
- https://github.com/kodadot/nft-gallery/issues/1839
- https://github.com/kodadot/nft-gallery/issues/2025
- https://github.com/kodadot/nft-gallery/issues/2049
- https://github.com/kodadot/nft-gallery/issues/1477
- https://github.com/kodadot/nft-gallery/issues/2100
- https://github.com/kodadot/nft-gallery/issues/2110

#### Analytics
- https://github.com/kodadot/nft-gallery/issues/1850
- https://github.com/kodadot/nft-gallery/issues/1285

#### Bugs
Picking something from category bugs will always cheer up our userbase and make their life way better experience
- https://github.com/kodadot/nft-gallery/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Abug

### Ref
- [Breakup of Yugoslavia](https://en.wikipedia.org/wiki/Breakup_of_Yugoslavia)

**Comment by vikiival** [comment]
**Viki's Meta Minutes**
-

### Cloudflare images:

the biggest update for the last two weeks is definitely Cloudflare images. 
Why? 

It brings a far lower data load on the page.
As an example the first page of donkeys downloads around 106 MB of data!
It is a lot of data to download and also process by the browser and the CPU.

<img width="1552" alt="Screenshot 2022-02-02 at 13 43 34" src="https://user-images.githubusercontent.com/22471030/152156218-0360a09f-5a7b-4a68-8877-5171fd1b68ce.png">

Now the same page using Cloudflare images need only 134 kB!!
*We save like 99.88%* of the traffic which is a huuuge number `(1 - (134 / 106 * 1024)) * 100`
<img width="1532" alt="Screenshot 2022-02-02 at 13 49 13" src="https://user-images.githubusercontent.com/22471030/152157026-604e3480-07d9-404b-b1a4-3053dedce2bb.png">

These images are served in the new AVIF format, developed by Netflix to give you the crispy image using as few bytes as possible.
What happens when we don't have a Cloudflare image?  We download an IPFS version out of it, so you will always get the image.

The new NFTs are downloaded from the indexer pushed to the image CDN every 30 minutes (we will cut the interval for 15 min).
Of course, it would not be possible without the community contribution from the @unknow112. He was able to tame the asynchronous python script so it's fast AF to process the images.

### Polishing UI:
Along with CF images we have other new features.
Now we are able to show **any** type of NFT media directly on the Gallery page.
This allows artists like *ClownWorldHouse* or *One Raw Artist* to have their 3D objects directly visible. 
We still recommend minting a poster with your 3D art so we can optimize the data load.

<img width="1017" alt="Screenshot 2022-02-02 at 14 01 11" src="https://user-images.githubusercontent.com/22471030/152158860-d8859377-44eb-47f3-915a-bd4dccf09958.png">

If you want to save some payload you can opt-out of this feature in the settings pane.

### Optimizing the app:
As @yangwao mentioned we did a ton of refactoring the maintain the highest code quality.
We refactored the minting UI and the Gallery item. As there is no visible output on the front-end, it brings a better developer experience and is easier to build new features, such as integration with other para chains.

Along with that @rchovatiya88 initialized end-to-end integration tests using cypress. What does it mean? Before we release the app the test will automatically click through the app and show the changes between the last snapshot and the PR. This will prevent some features will "disappear" from the app

### Finally: 
we have officially released the Statemine implementation of the KodaDot so feel free to mint some on-chain NFTs ❤️.

### Numbers:

Last weekend I played with the data analytics and found a couple of interesting graphs.

To this date, more than 40_000 KSM has been spent on the Kusama NFTs and we are really happy to be a community storefront for it.
<img width="895" alt="Screenshot 2022-01-30 at 11 53 12" src="https://user-images.githubusercontent.com/22471030/152162448-792fadb4-9ddf-4058-9ca6-ed3a9548e24a.png">

If we split the graph into the daily chart we see the daily spent amount of KSM. 
<img width="895" alt="Screenshot 2022-01-30 at 11 45 35" src="https://user-images.githubusercontent.com/22471030/152163027-d56d5f6e-00bd-462f-a71f-3d14d1fc9311.png">


### Packages

We moved our internal packages from `@vue-polkadot` to `@kodadot1`.

**Comment by damskyftw** [comment]
## Got something to share too

### Closed couple old issues and checked few boxes in #1619 - thanks to [docs](https://docs.kodadot.xyz/#docs-are-being-created)
- #529 
- #648
- #401 

### A lot of writing has to be done:

- NFT landing page for newcomers/people who don't even know what crypto is 
https://github.com/kodadot/docs/issues/2
- bug bounty program as wao mentioned
- RMTerra article on koda-forest and his efforts and collab with KodaDot - got materials from Pedro 
- Of course, ton of stuff for KodaDot doccumentation

### Discord

Work on membership screening has been done and gate was added

- [x] first part was to add a gate which should mitigate spammers - for know it's working
- second part is to make possibility for discord users to completely customize what they see by making a specific roles for specific categories - you can join the discussion and share your ideas/experience
- Whole rework of all the get-roles has to be done, i want to make one nice channel where all role making will be done, right now it's pretty scattered and not too good looking
- add og role and channel for heavy minters and early adopters

- Got rid of some low-traffic channels, colateral damage has been done, by deleting channel with 2 messages 🤣
- Discussion and work on this is continous⤵
- #2020 

- Some updates on onboarding for new users, continous work, gotta find some hidden bugs
 - In short: Membership screening was edited, message from carl bot disabled, want to work on this more, add some love to it and mainly find out why welcome screen isn't working lol. Ghosts everywhere
- #2061 

### Docs
And of course, the docs part. 
- was always looking to start "coding"
- missed some push or incentive to start doing it
- getting familiar with cmd, markdown, vuepress -  got a lot learning to do
- What always grows despite the markets is your experience
- Mainly working now on get-started stuff so we can easily point to this and people can learn
- also i've done tutorial on how to create polkadot.js wallet - currently waiting for some proofing and it can be uploaded to KodaDot yt channel which will be big part of docs - as you can see all the topics will and should be also in video form to satisfy all possible needs of new users
- the tutorial is watchable in #1445 
![ff](https://user-images.githubusercontent.com/92786519/152126734-f77e703c-a5c1-4277-bdd4-24fcea72a0dd.png)

Also, my time on beloved island in the middle of nowhere is shortening. Caffe will become my main office(curious on how this is gonna go lol) and i'll be in upcoming weeks heading back to europe, will finally have proper internet (ultra lol) and all the stuff i want to get to boost productivity! Good times ahead.

**Comment by yangwao** [comment]
I've looked on pull requests, seems we are finally **having more closed issues** than we are having opened, which is really a breakthrough! 

Kudos to all contributors keeping the pace!

![image](https://user-images.githubusercontent.com/5887929/152215359-30701a5d-491c-43eb-88a6-9cb59a1a31c4.png)
